store/tikv/prewrite.go

```diff
@@ -240,3 +283,25 @@ func (c *twoPhaseCommitter) prewriteMutations(bo *Backoffer, mutations Committer
 	// `doActionOnMutations` will unset `useOnePC` if the mutations is splitted into multiple batches.
 	return c.doActionOnMutations(bo, actionPrewrite{}, mutations)
 }
+
+// Reprewrite the primary lock to fallback from async commit or 1PC
+func (c *twoPhaseCommitter) reprewritePrimary(bo *Backoffer) {
+	if !atomic.CompareAndSwapUint32(&c.reprewrite.started, 0, 1) {
+		return
+	}
+	c.reprewrite.Add(1)
+	go func() {
+		// It is slow to find the primary mutation by iterating the mutations.
+		// But it is not a big problem because fallback happens rarely and the number of mutations are
+		// small in async commit or 1PC.
+		var mutation CommitterMutations
+		for i := 0; i < c.mutations.Len(); i++ {
+			if bytes.Equal(c.primary(), c.mutations.GetKey(i)) {
+				mutation = c.mutations.Slice(i, i+1)
+				break
+			}
+		}
+		c.reprewrite.err = errors.Trace(c.prewriteMutations(bo, mutation))
```

Comment:
Should we catch the panic error in this goroutine as it's may block the `execution` path?

Reply:
Is it also the reason that there is a `defer cancel` in https://github.com/pingcap/tidb/blob/051285a572/store/tikv/2pc.go#L1661? Then, maybe i would better not remove that cancel?

Reply:
I guess so seems we've no such test cases yet.

Reply:
I guess so, seems we've no such test cases yet.

Reply:
Panics are not expected to happen so there is no such a test I think. Anyway, I can add one for it.